pipeline/objects/environment/__init__.py

```diff
@@ -0,0 +1,70 @@
+import os
+from typing import List
+
+from pip._internal.commands.freeze import freeze
+
+"""
+TODO:
+1.  Add in dependency checks to check if a set of deps can be installed.
+    This is a solved problem and we can use the internal testing from pip:
+
+    from pip._internal.req.req_install import InstallRequirement
+    from pip._vendor.packaging.requirements import Requirement
+    from pip._internal.operations import check
+"""
+
+
+class Environment:
+    def __init__(
+        self,
+        environment_name: str = None,
+        dependencies: List[str] = [],
+        extra_index_urls: List[str] = [],
+        extend_environments: List["Environment"] = [],
+    ):
+        self.environment_name = environment_name
+        self.dependencies = dependencies
+        self.extra_index_urls = extra_index_urls
+        for _env in extend_environments:
+            self.merge_with_environment(_env)
+
+    def to_requirements(self, output_dir: str = "./") -> None:
+        requirements_path = os.path.join(os.path.join(output_dir, "requirements.txt"))
+        with open(requirements_path, "w") as req_file:
+            for _dep in self.dependencies:
+                req_file.write(f"{_dep}\n")
+
+    def add_dependency(self, dependency: str) -> None:
+        if self.initialized:
+            raise Exception(
+                "Cannot add dependency after the environment has \
+                been initialized."
+            )
+        self.dependencies.append(dependency)
+
+    def merge_with_environment(self, env: "Environment") -> None:
+        if not isinstance(env, Environment):
+            raise Exception("Can only merge with another environment")
+
+        self.dependencies.extend(env.dependencies)
```

Comment:
What happens for conflicting dependencies? e.g. if I merge an environment with `black==20.0.0` with an environment with `black==21.0.0`.

Reply:
At the moment there's no conflict checks. I believe the way that pip install from a `requirements.txt` the last most entry will be the one present in the final system.
This needs a lot more work, I think I'll just remove the merging for the moment actually.

Reply:
I'll remove the merging, it's not necessary right now.